base/channels.jl

```diff
@@ -0,0 +1,181 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+abstract ChannelData{T}
+
+type ChannelDataSingle{T} <: ChannelData{T}
+    full::Bool          # Need a separate flag here to handle the case where "nothing" is a valid value in "data"
+    data::Nullable{T}
+end
+
+type ChannelDataMultiple{T} <: ChannelData{T}
+    data::Array{T,1}
+    szp1::Int               # current channel size plus one
+    sz_max::Int             # maximum size of channel
+    take_pos::Int           # read position
+    put_pos::Int            # write position
+end
+
+type ChannelAddr
+    pid::Int
+    cid::Int
+end
+
+abstract AbstractChannel{T}
+
+type Channel{T, Many} <: AbstractChannel{T}
```

Comment:
It helps dispatch to the correct method for `put!`, `take!`, etc. For Channel sz 1, it is `false` and the backing store is a `Nullable{T}`. For channel sizes > 1, it is true and the backing store is an `Array{T, 1}`.